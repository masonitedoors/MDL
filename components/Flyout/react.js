import React from 'react'
import classNames from 'classnames/bind'
import PropTypes from 'prop-types'
import { React as ModalBackdrop } from 'components/ModalBackdrop'
import { React as Icon } from 'components/Icon'
import { mX } from '@masonite/svg-icons'
import styles from './style.module.scss'

const cx = classNames.bind(styles)

const Flyout = React.forwardRef(({
  children, onClose, heading, moreHeading, width = 477,
}, ref) => (
  <>
    <aside ref={ref} className={cx('aside')} style={{ width }}>
      <div className={cx('header')}>
        <div className={cx('header__left')}>
          <button type="button" onClick={onClose} className={cx('close')}>
            <Icon svg={mX} size={20} />
          </button>
          {heading && <h2>{heading}</h2>}
        </div>
        {moreHeading}
      </div>

      {children}
    </aside>
    <ModalBackdrop />
  </>
))

Flyout.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  onClose: PropTypes.func.isRequired,
  heading: PropTypes.string,
  moreHeading: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
}

Flyout.defaultProps = {
  heading: null,
  moreHeading: null,
}

export default Flyout
