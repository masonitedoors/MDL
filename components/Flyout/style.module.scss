@import "styles/index";

@keyframes slideInCategories {
  from { right: -100% }
  to { right: 0 }
}

.aside {
  display: flex;
  position: fixed;
  top: 0;
  right: -100%;
  z-index: 1001;
  height: 100%;
  flex-direction: column;
  background-color: $color-white;
  animation: slideInCategories ease-out .5s forwards;
}

.header {
  background-color: $color-gray;
  padding: $space-s;
  display: flex;
  align-items: center;
  justify-content: space-between;
  
  h2 {
    margin-bottom: 0;
  }

  &__left {
    display: flex;
  }
}

.close {
  position: relative;
  margin-right: $space-xs;
  padding: 0;
  border: none;
}
