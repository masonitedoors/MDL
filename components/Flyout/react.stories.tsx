import React, { useState } from "react";
import { Story, Meta } from "@storybook/react";
import { React as Button } from "../../components/Button";
import Flyout from "./react";

const Template = (args) => {
  const [show, setShow] = useState(false);
  const [messageShowing, setMessageShowing] = useState(false);

  const onClose = () => {
    setShow(false);
    setMessageShowing(true);
  };

  return (
    <>
      {show && (
        <Flyout {...args} onClose={onClose}>
          <div
            style={{
              border: "solid 1px green",
              width: 100,
              marginLeft: 100,
              marginTop: 100,
            }}
          >
            {"{children}"} show below the heading.
          </div>
        </Flyout>
      )}
      <label>
        Width:
        <input
          type="number"
          value={args.width}
          onChange={({ target }) => setWidth(+target.value)}
        />
      </label>
      <div>
        <Button onClick={() => setShow(true)}>Show Flyout</Button>
      </div>
      {messageShowing && (
        <p>
          The onClose just ran. The caller is responsible for actually closing.
        </p>
      )}
    </>
  );
};

export default {
  title: "Flyout",
  component: Flyout,
} as Meta;

export const Primary = Template.bind({});

Primary.args = {
  children: "Lorem ipsum",
  heading: "Lorem Ipsum Heading",
  moreHeading: "moreHeading",
  width: 477,
};
