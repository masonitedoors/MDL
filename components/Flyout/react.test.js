import React from 'react'
import {
  render, fireEvent,
} from '@testing-library/react'
import Flyout from './react'

describe('Flyout', () => {
  it('should render without crashing', () => {
    const { asFragment } = render(<Flyout />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with content', () => {
    const { asFragment } = render(
      <Flyout
        heading="My heading"
        moreHeading={<button>A test button</button>}
        width={700}
        className="some class-name"
        style={{ backgroundColor: "#abc" }}
        onMouseEnter={() => {}}
      >
        <div>Some body content</div>
      </Flyout>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should call onClose', () => {
    const onClose = jest.fn()
    render(<Flyout onClose={onClose} />)
    const closeButton = document.querySelector('.close')
    fireEvent.click(closeButton)
    expect(onClose).toHaveBeenCalled()
  })
})
