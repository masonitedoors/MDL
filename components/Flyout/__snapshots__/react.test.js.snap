// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flyout should render with content 1`] = `
<DocumentFragment>
  <aside
    class="aside some class-name"
    style="width: 700px; background-color: rgb(170, 187, 204);"
  >
    <div
      class="header"
    >
      <div
        class="header__left"
      >
        <button
          class="close"
          type="button"
        >
          <div
            class="icon "
            style="height: 20px; width: 20px;"
          >
            <svg
              version="1.1"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title />
              <path
                d="M13.4 12l5.3-5.3c0.4-0.4 0.4-1 0-1.4s-1-0.4-1.4 0l-5.3 5.3-5.3-5.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5.3 5.3-5.3 5.3c-0.4 0.4-0.4 1 0 1.4 0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l5.3-5.3 5.3 5.3c0.2 0.2 0.5 0.3 0.7 0.3s0.5-0.1 0.7-0.3c0.4-0.4 0.4-1 0-1.4l-5.3-5.3z"
              />
            </svg>
          </div>
        </button>
        <h2>
          My heading
        </h2>
      </div>
      <button>
        A test button
      </button>
    </div>
    <div>
      Some body content
    </div>
  </aside>
  <div
    class="modal-backdrop modal-backdrop--active"
  >
    <div
      class="modal-backdrop__content modal-backdrop__content-enter modal-backdrop__content-enter-active"
    />
  </div>
</DocumentFragment>
`;

exports[`Flyout should render without crashing 1`] = `
<DocumentFragment>
  <aside
    class="aside"
    style="width: 477px;"
  >
    <div
      class="header"
    >
      <div
        class="header__left"
      >
        <button
          class="close"
          type="button"
        >
          <div
            class="icon "
            style="height: 20px; width: 20px;"
          >
            <svg
              version="1.1"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title />
              <path
                d="M13.4 12l5.3-5.3c0.4-0.4 0.4-1 0-1.4s-1-0.4-1.4 0l-5.3 5.3-5.3-5.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5.3 5.3-5.3 5.3c-0.4 0.4-0.4 1 0 1.4 0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l5.3-5.3 5.3 5.3c0.2 0.2 0.5 0.3 0.7 0.3s0.5-0.1 0.7-0.3c0.4-0.4 0.4-1 0-1.4l-5.3-5.3z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </aside>
  <div
    class="modal-backdrop modal-backdrop--active"
  >
    <div
      class="modal-backdrop__content modal-backdrop__content-enter modal-backdrop__content-enter-active"
    />
  </div>
</DocumentFragment>
`;
